### DISCUSSION TITLE
Bias's not affected by backpropogation

### DISCUSSION THREAD

**Discussion by ShayCerny** [mrdbourke/pytorch-deep-learning#942] [Q&A] [answered]
While running my model I realized that only the weight is affected by training while the bias parameter remains the same. I went through the code to try and find an error but there was none I could see. Please help. here is the code:

Model:
```class LinearRegressionModel(
    nn.Module
):  # <- almost everything in pytorch inherits from nn
    def __init__(self):
        super().__init__()
        self.weight = nn.Parameter(
            torch.randn(1, requires_grad=True, dtype=torch.float)
        )

        self.bias = nn.Parameter(torch.randn(1, requires_grad=True, dtype=torch.float))

        # Forward method to define the computation in the model

    def forward(self, x: torch.Tensor) -> torch.Tensor:  # <- "x" is the input data
        return self.weight * x + bias  # this is the linear regression formula


torch.manual_seed(42)

model_0 = LinearRegressionModel()
```

Output of model.parameters():
```
[Parameter containing:
tensor([0.3367], requires_grad=True), Parameter containing:
tensor([0.1288], requires_grad=True)]
```

Loss and Optimizer:
`loss_fn = nn.L1Loss()`
`optimizer = torch.optim.SGD(model_0.parameters(), lr=0.01)`

Training:
```
epochs = 100

for epoch in range(epochs):
    # set the model to training mode
    model_0.train()  # sets the mode all parameters that require gradients to require them

    # 1. forward pass
    y_pred = model_0(X_train)

    # 2. Calculate loss
    loss = loss_fn(y_pred, y_train)

    # 3. opimizer zero grad
    optimizer.zero_grad()  # resets the gradients

    # 4. backpropogation
    loss.backward()

    # 5. Step the optimizer
    optimizer.step()

    ## Testing
    model_0.eval()  # turns off settings in the model not needed for evaluation/ testing

    with torch.inference_mode():  # turns off gradient tracking and a couple more things behind the scenes
        # 1. forward pass
        test_pred = model_0(X_test)
        # 2. calculate loss
        test_loss = loss_fn(test_pred, y_test)
    if epoch % 10 == 0:
        print(f"Epoch: {epoch} | Loss: {loss} | Test Loss: {test_loss}")
        print(model_0.state_dict())
```

Output from training:
```
Epoch: 0 | Loss: 0.14169074594974518 | Test Loss: 0.3198745846748352
OrderedDict([('weight', tensor([0.3406])), ('bias', tensor([0.1288]))])

Epoch: 10 | Loss: 0.12648078799247742 | Test Loss: 0.28516465425491333 
OrderedDict([('weight', tensor([0.3796])), ('bias', tensor([0.1288]))])

Epoch: 20 | Loss: 0.11127082258462906 | Test Loss: 0.25045472383499146 
OrderedDict([('weight', tensor([0.4186])), ('bias', tensor([0.1288]))])

Epoch: 30 | Loss: 0.0960608422756195 | Test Loss: 0.21574480831623077  
OrderedDict([('weight', tensor([0.4576])), ('bias', tensor([0.1288]))])

Epoch: 40 | Loss: 0.08085088431835175 | Test Loss: 0.1810348629951477  
OrderedDict([('weight', tensor([0.4966])), ('bias', tensor([0.1288]))])

Epoch: 50 | Loss: 0.06564091145992279 | Test Loss: 0.14632496237754822 
OrderedDict([('weight', tensor([0.5356])), ('bias', tensor([0.1288]))])

Epoch: 60 | Loss: 0.050430942326784134 | Test Loss: 0.11161501705646515
OrderedDict([('weight', tensor([0.5746])), ('bias', tensor([0.1288]))])

Epoch: 70 | Loss: 0.035220976918935776 | Test Loss: 0.07690508663654327 
OrderedDict([('weight', tensor([0.6136])), ('bias', tensor([0.1288]))]) 

Epoch: 80 | Loss: 0.02001100964844227 | Test Loss: 0.04219517856836319  
OrderedDict([('weight', tensor([0.6526])), ('bias', tensor([0.1288]))]) 

Epoch: 90 | Loss: 0.004801043774932623 | Test Loss: 0.007485252805054188
OrderedDict([('weight', tensor([0.6916])), ('bias', tensor([0.1288]))]) 

```

**Comment by ShayCerny** [comment, marked as answer]
For anyone who may end up here. I came back to this after a few hours and realized I had missed a 'self' in the forward function
